ENIGMAsystem/SHELL/Graphics_Systems/General/GSd3d.h

```diff
@@ -19,88 +19,119 @@
 #ifndef ENIGMA_GSD3D_H
 #define ENIGMA_GSD3D_H
 
+#include "GSblend.h"
+#include "GScolors.h"
 #include "Universal_System/scalar.h"
 #include <string>
 
-namespace enigma {
+namespace enigma_user {
 
-extern bool d3dMode;
-extern bool d3dHidden, d3dClipPlane, d3dZWriteEnable;
-extern bool d3dPerspective;
-extern bool d3dLighting;
-extern bool d3dShading;
-extern int d3dCulling, d3dDepthOperator;
+// ***** RENDER STATE CONSTANTS *****
+enum {
+  rs_nicest,
+  rs_fastest,
+  rs_dontcare,
+};
 
-extern bool d3dFogEnabled;
-extern int d3dFogColor, d3dFogMode, d3dFogHint;
-extern float d3dFogStart, d3dFogEnd, d3dFogDensity;
+enum {
+  rs_exp,    // D3DFOG_EXP
+  rs_exp2,   // D3DFOG_EXP2
+  rs_linear, // D3DFOG_LINEAR
+};
+
+enum {
+  rs_never, 	// Always False            D3DCMP_NEVER            GL_NEVER
+  rs_less, 	    // source Z < depth Z      D3DCMP_LESS             GL_LESS
+  rs_equal,	    // source Z = depth Z      D3DCMP_EQUAL            GL_EQUAL
+  rs_lequal, 	// source Z <= depth Z     D3DCMP_LESSEQUAL        GL_LEQUAL
+  rs_greater,	// source Z > depth Z      D3DCMP_GREATER          GL_GREATER
+  rs_notequal, 	// source Z != depth Z     D3DCMP_NOTEQUAL         GL_NOTEQUAL
+  rs_gequal, 	// source Z >= depth Z     D3DCMP_GREATEREQUAL     GL_GEQUAL
+  rs_always     // Always True             D3DCMP_ALWAYS           GL_ALWAYS
+};
+
+enum {
+  rs_keep,   // GL_KEEP
+  rs_zero,      //  GL_ZERO
+  rs_replace,     //  GL_REPLACE
+  rs_incr,  // GL_INCR
+  rs_incr_wrap, // GL_INCR_WRAP
+  rs_decr,  // GL_DECR
+  rs_decr_wrap,  // GL_DECR_WRAP
+  rs_invert     // GL_INVERT
+};
+
+// NOTE: Game Maker uses clockwise culling to define the front face,
+// OpenGL's mode sets what defines the front face, Direct3D's mode sets what defines the back face
+enum {
+  rs_none, // No culling
+  rs_cw, // Clockwise culling
+  rs_ccw // Counter-clockwise culling
+};
+
+enum {
+  rs_back,
+  rs_front,
+  rs_front_back
+};
+
+enum {
+  rs_point, // Render vertices as points
+  rs_line,  // Render in wireframe mode
+  rs_solid  // Normal render mode
+};
+
+} // namespace enigma_user
+
+namespace enigma {
+
+struct Sampler {
+  int texture=-1; // GML texture id, NOT GL texture id!
+  bool wrapu=false, wrapv=false, wrapw=false;
+  bool interpolate=false;
+};
 
 struct Light {
   gs_scalar x=0, y=0, z=0, range=0;
   bool directional=false;
   int color=0;
 };
 
-extern int d3dLightsActive, d3dLightingAmbient;
-const Light& get_active_light(int id);
+struct RenderState {
```

Comment:
On that note... How big is this struct? We may want to rearrange it to keep it within a page.

Reply:
![RenderState struct Size in Visual Studio](https://user-images.githubusercontent.com/3212801/58751203-8fc51080-8469-11e9-9b8e-93071e45c877.png)

Reply:
Seems fine. Carry on. Note that if this gets out of hands, you can also reduce the size by changing the order of the elements in this struct. You're wasting a byte here and there by declaring three bools followed by a short, etc.

Reply:
I never knew, thanks for that.